### PR TITLE
expression: fix date_add interval month diffs from mysql

### PR DESCRIPTION
**Pull request by haplone**
### What problem does this PR solve? 

date_add interval month diffs from mysql

when we execute select date_add('2018-01-31',interval 1 month) in mysql we got 2018-02-28

but in tidb we got 2018-03-03

### What is changed and how it works?

1. fix the gap between golang api time.DateDate and mysql date_sub with days in expression/builtin_time.go
2. add GetLastDay in types/mytime.go to get last day in month
3. add test for above methods

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8534)

### CODE REVIEW

**Comment by haplone** [comment]
@shenli PTAL

File: expression/builtin_time.go [hunk 1]
```diff
@@ -2585,6 +2585,42 @@ func (du *baseDateArithmitical) getIntervalFromInt(ctx sessionctx.Context, args
...
+		if r4 == 0 && (r100 != 0 || r400 == 0) {
+			day = 29
+		}
+	}
+
+	return day
+}
+
+func getFixDays(year, month, day int, o time.Time) int {
+	if year == 0 && month != 0 && day == 0 {
```
Comment (zz-jason) [inline expression/builtin_time.go]: Looks like we should also handle the leap year case, in MySQL:
```sql
MySQL(root@localhost:test) > select date_add('2000-02-29',interval 1 year);
+----------------------------------------+
| date_add('2000-02-29', interval 1 year) |
+----------------------------------------+
| 2001-02-28                             |
+----------------------------------------+
1 row in set (0.01 sec)
```

File: expression/builtin_time.go [hunk 2]
```diff
@@ -2585,6 +2585,42 @@ func (du *baseDateArithmitical) getIntervalFromInt(ctx sessionctx.Context, args
 	return strconv.FormatInt(interval, 10), false, nil
 }
 
+var daysByMonth = [12]int{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
+
+func getMaxDayInMonth(year, month int) int {
```
Comment (zz-jason) [inline expression/builtin_time.go]: this function duplicates the functionality of `builtinLastDaySig.evalTime()`, can we extract a common function and make them share the same code logic?

Comment (haplone) [reply expression/builtin_time.go]: how about I extract getMaxDayInMonth to types/MysqlTime ,just like IsLeapYear?
@zz-jason

Comment (zz-jason) [reply expression/builtin_time.go]: OK. we can discuss the function name later.

Comment (haplone) [reply expression/builtin_time.go]: how about MaxDayInMonth?
I have another problem, `year, month, day` in MysqlTime is private, and I could not instance it in builtin_time.go ,now we only use it in types/Time ,could I add a func to instance MyqlTime , or I could only use MysqlTime by Time ?

Comment (haplone) [reply expression/builtin_time.go]: @zz-jason

Comment (zz-jason) [reply expression/builtin_time.go]: I think we only need "year" and "month" to get the last day of that month, so we can abstract a function like `GetLastDay()`:
```go
func getLastDay(year, month int) {
	...
}
```

And in `builtinLastDaySig.evalTime()`, after we have get the input time, we can further get the year and month info, then call this function to get the last day.

In `getFixDays()` we can call this function as well in order to get the last day.

File: types/mytime.go [hunk 3]
```diff
@@ -119,7 +119,25 @@ func (t MysqlTime) GoTime(loc *gotime.Location) (gotime.Time, error) {
...
+func isLeapYear(year uint16) bool {
+	return (year%4 == 0 && year%100 != 0) || year%400 == 0
+}
+
+var daysByMonth = [12]int{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
+
+// GetLastDay returns the last day of the month
+func GetLastDay(year, month int) int {
+	var day = 0
+	if month > 0 {
```
Comment (zz-jason) [inline types/mytime.go]: should we also check `month <= 12`?

Comment (haplone) [reply types/mytime.go]: I've modified , and moved this PR to [PR8988](https://github.com/pingcap/tidb/pull/8988) (merge into master first ), as shenli suggested .
review at PR8988 ,and help me close this PR. thanks!

**Comment by zz-jason** [comment]
@haplone https://github.com/pingcap/tidb/pull/8988 was merged, can we close this PR now?

**Comment by haplone** [comment]
@zz-jason Close it, thanks

**Pull request closed without merging**